cluster_management/src/main/java/com/pinterest/rocksplicator/Spectator.java

```diff
@@ -130,15 +130,30 @@ public static void main(String[] args) throws Exception {
 
     CuratorFramework zkClient = CuratorFrameworkFactory.newClient(zkConnectString, new ExponentialBackoffRetry(1000, 3));
     zkClient.start();
+
     InterProcessMutex mutex = new InterProcessMutex(zkClient, getClusterLockPath(clusterName));
+    LOG.error("Trying to obtain lock");
+
+    /**
+     * TODO: Improve this logic.
+     *
+     * This is an issue if the lock is released without the process going down. This can happen,
+     * if the client looses connection to zookeeper for a period of time. In such a scenario,
+     * the client doesn't have a way to stop spectator from processing shard_map generation.
+     * The lock may subsequently be obtained by another instance of spectator and hence breaking
+     * the semantics of single updator.
+     * In that case, the spectator will continue to listen and generate shard_map.
```

Comment:
Guodong is working on a fix for this via a connection state listener interface.

Reply:
As. a long term solution, I would prefer a distributed controller approach and have Spectator instance per cluster be managed through LeaderStandby state model.

Reply:
we can treat the connection state listener as a short term fix

Reply:
As a long term solution: https://github.com/pinterest/rocksplicator/pull/412/ is a first step towards distributed spectator at par. with distributed controller.